File: backend/app/models/estate.py
from dataclasses import dataclass, field
from typing import List, Optional


@dataclass
class KpiModel:
    key: str
    label: str
    budget: float
    actual: float
    achieved: Optional[float] = None
    gsa: Optional[float] = None
    rank: Optional[int] = None


@dataclass
class CropSectionModel:
    crop_name: str
    kpis: List[KpiModel] = field(default_factory=list)


@dataclass
class EstateModel:
    id: str
    key: str
    name: str
    order: int
    row_type: str
    status: str
    crop_variance: float
    profit_actual: float

    crop_budget: float = 0.0
    crop_actual: float = 0.0

    nsa_budget: float = 0.0
    nsa_actual: float = 0.0

    cop_budget: float = 0.0
    cop_actual: float = 0.0

    profit_budget: float = 0.0

    kpis: List[KpiModel] = field(default_factory=list)
    crops: List[CropSectionModel] = field(default_factory=list)

File: frontend/src/types/estate.ts
export type Status = "red" | "orange" | "green";
export type RowType = "operational" | "group" | "total";
export type EstateGroup = "HIGH GROWN" | "LOW GROWN" | "REPROCESS" | "CINNAMON";

export type Kpi = {
  key: string;
  label: string;
  budget: number;
  actual: number;
  achieved?: number | null;
  previous_year_budget?: number | null;
  previous_year_actual?: number | null;
  gsa?: number | null;
  rank?: number | null;
};

export type PeriodLabels = {
  prev_budget: string;
  prev_actual: string;
  budget: string;
  actual: string;
};

export type EstateHistoryPoint = {
  month: string;
  crop_budget: number;
  crop_actual: number;
  nsa_budget: number;
  nsa_actual: number;
  cop_budget: number;
  cop_actual: number;
  profit_budget: number;
  profit_actual: number;
};

export type CropSection = {
  crop_name: string;
  kpis: Kpi[];
};

export type EstateListItem = {
  id: string;
  key: string;
  name: string;
  order: number;
  row_type: RowType;
  status: Status;
  crop_variance: number;
  profit_actual: number;
  crop_budget?: number;
  crop_actual?: number;
  nsa_budget?: number;
  nsa_actual?: number;
  cop_budget?: number;
  cop_actual?: number;
  profit_budget?: number;
};

export type EstateDetail = EstateListItem & {
  kpis: Kpi[];
  crops: CropSection[];
  history?: EstateHistoryPoint[];
  period_labels?: PeriodLabels | null;
};